src/Windows/dll/local_cache.cpp

```diff
@@ -1,33 +1,412 @@
+        hAlg,
+        &hHash,
+        pbHashObject,
+        cbHashObject,
+        NULL,
+        0,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptCreateHash\n";
+        goto Cleanup;
+    }
+
+
+    //hash some data
+    if (!NT_SUCCESS(status = BCryptHashData(
+        hHash,
+        (PBYTE)data,
+        (ULONG)data_size,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptHashData\n";
+        goto Cleanup;
+    }
+
+    //close the hash
+    if (!NT_SUCCESS(status = BCryptFinishHash(
+        hHash,
+        pbHash,
+        cbHash,
+        0)))
+    {
+        errorString = "Error 0x" + std::to_string(status) + "returned by BCryptFinishHash\n";
+        goto Cleanup;
+    }
+
+    retval.reserve(2 * cbHash + 1);
+    for (size_t i = 0; i < cbHash; i++)
+    {
+        char buf[3];
+        snprintf(buf, sizeof(buf), "%02x", pbHash[i]);
+        retval += buf;
+    }
+
+Cleanup:
+
+    if (hAlg)
+    {
+        BCryptCloseAlgorithmProvider(hAlg, 0);
+    }
+
+    if (hHash)
+    {
+        BCryptDestroyHash(hHash);
+    }
+
+    if (pbHashObject)
+    {
+        HeapFree(GetProcessHeap(), 0, pbHashObject);
+    }
+
+    if (pbHash)
+    {
+        HeapFree(GetProcessHeap(), 0, pbHash);
+    }
+
+    throw_if(!NT_SUCCESS(status), errorString);
+
+    return retval;
+}
+
+static std::string sha256(const std::string& input)
+{
+    return sha256(input.length(), input.data());
+}
+
+static std::string get_file_name(const std::string& id)
+{
+    return g_cache_dirname + "\\" + sha256(id);
+}
+
+void local_cache_clear() {
+    init();
+
+    WIN32_FIND_DATA data;
+    std::wstring baseDir(g_cache_dirname.begin(), g_cache_dirname.end());
+    std::wstring searchPattern = baseDir + L"\\*";
+
+    HANDLE hFind = FindFirstFile(searchPattern.c_str(), &data);
+
+    if (hFind != INVALID_HANDLE_VALUE)
+    {
+        do {
+            std::wstring fileName(data.cFileName);
+            if (fileName.compare(L".") != 0 && fileName.compare(L"..") != 0)
+            {
+                std::wstring fullFileName = baseDir + L"\\" + fileName;
+
+                throw_if(!DeleteFileW(fullFileName.c_str()),
+                    "Deleting file failed, error code " + GetLastError());
+            }
+        } while (FindNextFile(hFind, &data));
+        FindClose(hFind);
+    }
+
+    return;
+};
+
+HANDLE OpenHandle(LPCWSTR lpFileName, DWORD dwDesiredAccess, DWORD dwShareMode, LPSECURITY_ATTRIBUTES lpSecurityAttributes,
+    DWORD dwCreationDisposition, DWORD dwFlagsAndAttributes, HANDLE hTemplateFile) {
+    
+    HANDLE file;
+    int i = 0;
+    do {
+        file = CreateFile(lpFileName, dwDesiredAccess, dwShareMode, lpSecurityAttributes,
+            dwCreationDisposition, dwFlagsAndAttributes, hTemplateFile);
+        
+        Sleep(SLEEP_RETRY_MS);
+        i++;
+    } while ((file == INVALID_HANDLE_VALUE) &&
+        (GetLastError() == 32) &&
```

Comment:
Don't use the manifest constant of 32, instead define ERROR_SHARING_VIOLATION as 32 and use that.

Reply:
Done